src/DomainGuest.sol

```diff
@@ -0,0 +1,374 @@
+        grain = limit;
+
+        payload = abi.encodeWithSelector(DomainHost.release.selector, rid++, burned);
+
+        emit Release(burned);
+    }
+
+    /// @notice Push surplus (or deficit) to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function _push() internal isLive returns (bytes memory payload) {
+        uint256 _dai = vat.dai(address(this));
+        uint256 _sin = vat.sin(address(this));
+        if (_dai >= _sin + dust) {
+            // We have a surplus
+            if (_sin > 0) vat.heal(_sin);
+
+            uint256 wad = (_dai - _sin) / RAY;    // Leave the dust
+
+            // Burn the DAI and unload on the other side
+            vat.swell(address(this), -_int256(wad * RAY));
+            payload = abi.encodeWithSelector(DomainHost.push.selector, rid++, _int256(wad));
+
+            emit Push(int256(wad));
+        } else if (_sin >= _dai + dust) {
+            // We have a deficit
+            if (_dai > 0) vat.heal(_dai);
+
+            int256 deficit = -_int256(_divup(_sin - _dai, RAY));    // Round up to overcharge for deficit
+            payload = abi.encodeWithSelector(DomainHost.push.selector, rid++, deficit);
+
+            emit Push(deficit);
+        } else {
+            revert("DomainGuest/dust");
+        }
+    }
+
+    /// @notice Merge DAI into surplus
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI that has been sent to this domain [WAD]
+    function rectify(uint256 _lid, uint256 wad) external hostOnly ordered(_lid) {
+        vat.swell(address(this), _int256(wad * RAY));
+
+        emit Rectify(wad);
+    }
+
+    /// @notice Trigger the end module
+    /// @param _lid Local ordering id
+    function cage(uint256 _lid) external hostOnly isLive ordered(_lid) {
+        live = 0;
+        end.cage();
+
+        emit Cage();
+    }
+
+    /// @notice Set the cure value for the host
+    /// @dev Triggered during shutdown
+    function _tell() internal returns (bytes memory payload) {
+        uint256 debt = end.debt();
+        require(debt > 0 || (vat.debt() == 0 && live == 0), "DomainGuest/end-debt-zero");
+        uint256 _grain = grain * RAY;
+        uint256 cure = _grain > debt ? _grain - debt : 0;
+
+        payload = abi.encodeWithSelector(DomainHost.tell.selector, rid++, cure);
+
+        emit Tell(cure);
+    }
+
+    /// @notice Mint a claim token for the given user
+    /// @dev    Claim amount is in units of local debt.
+    /// @param usr The destination to send the claim tokens to
+    /// @param wad The amount of claim tokens to mint
+    function exit(address usr, uint256 wad) external hostOnly {
+        claimToken.mint(usr, wad);
+
+        emit Exit(usr, wad);
+    }
+
+    function heal(uint256 amount) external {
+        vat.heal(amount);
+    }
+    function heal() external {
+        vat.heal(_min(vat.dai(address(this)), vat.sin(address(this))));
+    }
+
+    // --- Canonical DAI Support ---
```

Comment:
Super subjective, but as this is the simplest path and common task, we might want to put the canonical dai support in the beginning of the file, then have MCD, and at the end Teleport.

Reply:
(same for Host)